dxr/app.py

```diff
@@ -315,36 +333,104 @@ def sidebar_links(sections):
               url_for('.parallel', tree=t['name'], path=path),
               t['description'])
             for t in frozen_configs()],
-        'generated_date': frozen['generated_date'],
+        'generated_date': date,
         'google_analytics_key': config.google_analytics_key,
         'filters': filter_menu_items(
             plugins_named(frozen_config(tree)['enabled_plugins'])),
-    }
-
-    # File template variables
-    file_vars = {
+        # File template variables
         'paths_and_names': _linked_pathname(path, tree),
         'icon': icon(path),
         'path': path,
-        'name': basename(path),
+        'name': basename(path)
     }
 
+
+def _browse_file(tree, path, line_docs, file_doc, config, date = None, contents = None):
+    """Return a rendered page displaying a source file.
+
+    :arg string tree: name of tree on which file is found
+    :arg string path: relative path from tree root of file
+    :arg list line_docs: LINE documents as defined in the mapping of core.py,
+        where the `content` field is dereferenced
+    :arg file_doc: the FILE document as defined in core.py
+    :arg config: TreeConfig object of this tree
+    :arg date: a formatted string representing the generated date, default to now
+    :arg string contents: the contents of the source file, defaults to joining
+        the `content` field of all line_docs
+    """
+    def sidebar_links(sections):
+        """Return data structure to build nav sidebar from. ::
+
+            [('Section Name', [{'icon': ..., 'title': ..., 'href': ...}])]
+
+        """
+        # Sort by order, resolving ties by section name:
+        return sorted(sections, key=lambda section: (section['order'],
+                                                     section['heading']))
+
+    if not date:
+        # Then assume that the file is generated now. Remark: we can't use this
+        # as the default param because that is only evaluated once, so the same
+        # time would always be used.
+        date = datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S +0000")
+
+    common = _build_common_file_template(tree, path, date, config)
+    links = file_doc.get('links', [])
     if is_image(path):
         return render_template(
             'image_file.html',
-            **merge(common, file_vars))
+            **common)
     else:  # For now, we don't index binary files, so this is always a text one
```

Comment:
I change this to "We don't allow browsing binary files, so this must be a text file," since we gray out links to binary files and don't allow clicking them. (however maybe the change is more suitable for PR 421).